### ISSUE TITLE
It shouldn't be possible to create instances that cannot be opened

### ISSUE THREAD

**Issue by evlist** [orbeon/orbeon-forms#2157]
With Form Runner / Form Builder it is quite easy to store instances which cannot be opened again.

For instance:

* Create a form with 2 integer fields (a and b) and a float field which is calculated as "$a div $b".
* Deploy this form and create a new instance
* Enter 0 in b
* Close the error dialog
* Save
* Go to summary
* The instance that you've just created cannot be opened again (you get an error page)

And if you are using versioning you can't even modify the form to suppress the error since your instance is bound to the form version where the error happens.

There are two ways to fix this issue:

* Disabling save when XPath errors are detected
* Find a way to enable loading instances that cause XPath errors.

The second option would be much more useful from a user perspective.

**Comment by ebruchez** [comment]
By default, we consider some errors fatal upon form load. The idea was that if that was happening, it was probably an error in an expression. But this case, and others, show that it's not the case, so this might be a bad idea outside of development, if not always.

With this property this behavior can be disabled:

```xml
<property
  as="xs:boolean"
  name="oxf.xforms.fatal-errors-during-initialization"
  value="false"/>
```

**Comment by evlist** [comment]
I was not aware of this property which seems very useful, thanks!

I think it would be a good idea to set it to false, at least in dev mode but my point was also that it should be impossible to create situations that are difficult to "unlock". What about preventing to save instances that generate these errors (this behavior could be controlled either by this property or by another one)?

Also, even if should generally be avoided to give technical details in error messages in production, in the case of XPath errors by user entry do not show that much information:

![capture du 2015-04-02 08 08 08](https://cloud.githubusercontent.com/assets/109022/6958700/d1509114-d90f-11e4-9453-ce3d499b7a54.png)

Formating such errors in more user friendly FR dialogs (saying something such as "Your latest entry has caused an integer division by zero") would be very helpful for end users to correct their errors by themselves.

**Comment by ebruchez** [comment]
I think it depends what we want to allow users to write as expressions.

We had decided that XPath type errors would be non-fatal. For example: `$a + $b` and `$a` is empty? A type error is thrown. Of course you could write `if ($a castable as xs:integer)` etc. but that is a lot of boilerplate.

What about `$a div $b`? Arguably the case is similar: the expression doesn't fail with a type error, but it fails for similar reasons: the data entered by the user (or via other calculations) does not make sense for the given expression. So the calculation fails, and the result of the calculation is set to blank.

This idea argues in favor of making all these dynamic errors non-fatal.

Now consider a form again with a calculated field `$a + $b`: should we allow saving that form if `$a` was empty? Probably yes if `$a` is optional, and if the result of the calculation can be empty too! So I am not sure that we can prevent saving the form in the general case. What should matter for saving the form is data validity, at least that's the logic so far.

Could we say that some given calculations *must* succeed for the form to be valid? But when in what situation would this arise which wouldn't be caught by the rest of form validation? For example, if a field cannot be blank, or a divisor cannot be `0`, a validation should enforce this.

BTW on the error dialog: by default in prod mode we don't show any error message at all. These errors are not meant to be end-user-facing, but developer-facing.

In short, I am not sure what's the best solution to improve the current behavior, except that we should set `oxf.xforms.fatal-errors-during-initialization` to `false` by default so that form data that can be saved can also be reloaded.

**Comment by evlist** [comment]
That's a complex issue indeed!

Thanks,

Eric
